### PR TITLE
feat: enable Windows Control Overlay on Linux

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/23665.

This PR enables the [Windows Control Overlay API](https://developer.mozilla.org/en-US/docs/Web/API/Window_Controls_Overlay_API) on Linux. To do so, we create a new subclass of `FramelessView`. This primarily follows logic from [upstream's implementation](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/views/frame/opaque_browser_frame_view.cc) with modifications for our use case and stripped of logic irrelevant to Electron.

cc @bpasero @deepak1556 

Todo:
- [x] Ensure existing WCO tests pass
- [x] Test with RTL and LTR
- [x] Test that color can be changed dynamically
- [x] Test that height can be changed dynamically

Functionality on Wayland will be approached as a follow-up as I don't currently have the requisite hardware to test this.

<details><summary>WCO RTL</summary>
<p>

Screenshot 2024-04-03 at 9 28 20 PM

</p>
</details> 

<details><summary>WCO LTR</summary>
<p>

Screenshot 2024-04-03 at 9 29 06 PM

</p>
</details> 

<details><summary>Dynamic WCO Update</summary>
<p>

Gif Demonstration with the following code:

```js
const { app, BrowserWindow } = require('electron');

async function createWindow () {
  const mainWindow = new BrowserWindow({
    titleBarOverlay: {
      color: '#0000f0',
      symbolColor: '#f54260'
    },
    titleBarStyle: 'hidden'
  });

  await mainWindow.loadURL('data:text/html,<h1>Hello, World!</h1>');

  setTimeout(() => {
    console.log('Setting title bar overlay color!!!');
    mainWindow.setTitleBarOverlay({
      color: '#355E3B',
      symbolColor: '#99aebd',
      height: 40
    });
  }, 3000);
}

app.whenReady().then(() => {
  createWindow();

  app.on('activate', () => {
    if (BrowserWindow.getAllWindows().length === 0) createWindow();
  });
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') app.quit();
});
```

wco-update-linux

</p>
</details> 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Enabled the Windows Control Overlay API on Linux.

### CODE REVIEW

File: docs/api/structures/base-window-options.md
Comment (erickzhao) [inline docs/api/structures/base-window-options.md:90]:
```suggestion
  * `customButtonsOnHover` _macOS_ - Results in a hidden
```

**Review by erickzhao — APPROVED** [review]
API LGTM

**Review by jkleinsc — COMMENTED** [review]
API LGTM

**Comment by deepak1556** [comment]
Sorry for the delay here, need some time to verify the changes. I will have an update later this week.

**Comment by ckerr** [comment]
API LGTM

File: shell/browser/ui/views/opaque_frame_view.h
Comment (deepak1556) [inline shell/browser/ui/views/opaque_frame_view.h:38]: This causes a think frame border for resizing outside the client area in window restored mode, also this seems chrome UI specific. Can we remove it ?

**Review by deepak1556 — COMMENTED** [review]
Works great, left feedback on the frame border.

**Comment by deepak1556** [comment]
There is also the issue with theme adjustment with the current implementation,

**native titlebar:**
Screenshot 2024-04-24 at 17 21 22

**custom titlebar:**

Screenshot 2024-04-24 at 17 21 35

I believe the issue arises from the class used to draw the caption buttons `//ui/views/window/frame_caption_button.h` which does not use any gtk api but rather has a generic Chrome specific implementation. On the other hand, `ClientFrameViewLinux` uses `//ui/gtk/nav_button_provider_gtk` which seems to work well across different themes. Is there a reason we cannot reuse `ClientFrameViewLinux` to support WCO on x11 ?

@ruihe774  since you have been working on this support for wayland, do you have thoughts on the above ? My question is mainly if the window caption container code can be shared between wayland and x11.

**Comment by ruihe774** [comment]
> @ruihe774 since you have been working on this support for wayland, do you have thoughts on the above ? My question is mainly if the window caption container code can be shared between wayland and x11.

Yes. I've also used `ClientFrameViewLinux` to implement WCO for x11 in 426f5e1a4ddee4989da999e77ef2661d911d7ab8. The title of my PR has not updated yet, which is misleading. Sorry for that.

**Comment by deepak1556** [comment]
Missed that, thanks for clarifying! If that works well for both x11 and wayland then seems like a better route. @codebytere thoughts ?

**Comment by ckerr** [comment]
I would mostly be OK with either this or 41840. Both PRs have strengths and weaknesses; so if any other maintainer has strong feelings about which to use, TBH I could go along with either.

If we go with this PR, I think I'm OK with it once the merge conflicts and `kFrameBorderThickness` have been resolved.

It would be nice-to-have to see better theming as per https://github.com/electron/electron/pull/41769#issuecomment-2074460912 but don't see it as a blocker for this PR. I would be OK with addressing that in a follow-up.

**Comment by ckerr** [comment]
I would mostly be OK with either this or 41840. Both PRs have strengths and weaknesses; so if any other maintainer has strong feelings about which to use, TBH I could go along with either.

The only blockers on this PR IMO are the merge conflicts and `kFrameBorderThickness`. It would be nice-to-have to see better theming as per https://github.com/electron/electron/pull/41769#issuecomment-2074460912 but don't see it as a blocker for this PR. I would be OK with addressing that in a follow-up.

File: shell/browser/api/electron_api_base_window.cc
Comment (deepak1556) [inline shell/browser/api/electron_api_base_window.cc:1049]: Method declaration in `BaseWindow::BuildPrototype` is still behind `IS_WIN`

**Review by deepak1556 — COMMENTED** [review]
I am good with aligning the implementation with Chrome. My original ask to try GTK rendering was to match the native titlebar but given Chrome doesn't use any GTK for their custom titlebar either makes sense to follow the route which has been tested and worked well in upstream.

**Review by deepak1556 — APPROVED** [review]

**Review by ckerr — APPROVED** [review]

**Pull request merged**
